### PR TITLE
follow up to #36294: allow migrate sub stream state with custom partition router

### PR DESCRIPTION
**Pull request by lazebnyi**
## What

The implementation proposed in [this PR](https://github.com/airbytehq/airbyte/pull/36294) allows migrating sub-stream state only if the stream uses `SubstreamPartitionRouter` as the partition router. However, for some streams that use `CustomPartitionRouter` instead of the default, it would also be beneficial to have a solution for migration (e.g. Source Survey Monkey - `survey_responses` stream).

## How

- Update type validation in `create_legacy_to_per_partition_state_migration`.
- Add conditional retrieval for `parent_key` in the initialization logic of `LegacyToPerPartitionStateMigration`.

### CODE REVIEW

File: airbyte-cdk/python/airbyte_cdk/sources/declarative/parsers/model_to_component_factory.py [hunk 1]
```diff
@@ -325,7 +325,7 @@ def create_legacy_to_per_partition_state_migration(
             raise ValueError(
                 f"LegacyToPerPartitionStateMigrations can only be applied on a DeclarativeStream with a SimpleRetriever. Got {type(declarative_stream.retriever)}"
             )
-        if not isinstance(declarative_stream.retriever.partition_router, SubstreamPartitionRouterModel):
+        if not isinstance(declarative_stream.retriever.partition_router, (SubstreamPartitionRouterModel, CustomPartitionRouterModel)):
```
Comment (girarda) [inline airbyte-cdk/python/airbyte_cdk/sources/declarative/parsers/model_to_component_factory.py:328]: this won't work for any arbitrary CustomPartitionRouter, right?

there's a few more requirements
- It needs to be a substream partition router
- its parent_stream_config must be a dict

Is that right? can we encode this somehow?

Comment (lazebnyi) [reply airbyte-cdk/python/airbyte_cdk/sources/declarative/parsers/model_to_component_factory.py:328]: I believe `parent_stream_config` always dict if we use `CustomPartitionRouter` as it parsed in `create_custom_component`

Comment (girarda) [reply airbyte-cdk/python/airbyte_cdk/sources/declarative/parsers/model_to_component_factory.py:328]: you're correct on the typing, but not all CustomPartitionRouter necessarily have a parent_stream_config.

We should at least verify its existence.

Fine to keep out of scope here, but this makes me think we're missing an abstraction. We should introduce CustomSubstreamPartitionRouters so we can enforce the presence of the parent_stream_config.

File: airbyte-cdk/python/unit_tests/sources/declarative/migrations/test_legacy_to_per_partition_migration.py [hunk 2]
```diff
@@ -241,3 +242,33 @@ def _migrator_with_multiple_parent_streams():
     config = {}
     parameters = {}
     return LegacyToPerPartitionStateMigration(partition_router, cursor, config, parameters)
+
+
+def _migrator_with_custom_partition_router():
```
Comment (girarda) [inline airbyte-cdk/python/unit_tests/sources/declarative/migrations/test_legacy_to_per_partition_migration.py:247]: is this used? we should probably add 2 unit tests
1. with a custom partition router that has parent_stream_configs
2. with a custom partition router that does not have a parent_stream_configs

File: airbyte-cdk/python/airbyte_cdk/sources/declarative/parsers/model_to_component_factory.py [hunk 3]
```diff
@@ -325,10 +325,20 @@ def create_legacy_to_per_partition_state_migration(
             raise ValueError(
                 f"LegacyToPerPartitionStateMigrations can only be applied on a DeclarativeStream with a SimpleRetriever. Got {type(declarative_stream.retriever)}"
             )
-        if not isinstance(declarative_stream.retriever.partition_router, SubstreamPartitionRouterModel):
+
+        partition_router = declarative_stream.retriever.partition_router
+        router_class = partition_router[0].get("class_name") if partition_router else None
+
+        is_default_substream_partition_router = isinstance(partition_router, SubstreamPartitionRouterModel)
+        is_custom_substream_partition_router = router_class == SubstreamPartitionRouterModel and isinstance(
```
Comment (girarda) [inline airbyte-cdk/python/airbyte_cdk/sources/declarative/parsers/model_to_component_factory.py:333]: I'm not sure I get this check. would the router_class also be of type `SubstreamPartitionRouterModel` when it's custom?

I think we just need to check that the custom class has an attribute `parent_stream_config` as a proxy for verifying it'll work with the migration

Comment (lazebnyi) [reply airbyte-cdk/python/airbyte_cdk/sources/declarative/parsers/model_to_component_factory.py]: But `router_class` can be `ListPartitionRouter` as well, so I believe we need check `router_class` type

Added `is_parent_stream_config` validation

File: airbyte-cdk/python/unit_tests/sources/declarative/migrations/test_legacy_to_per_partition_migration.py (hunk 2 above)
Comment (lazebnyi) [reply airbyte-cdk/python/unit_tests/sources/declarative/migrations/test_legacy_to_per_partition_migration.py]: Added

**Review by girarda — APPROVED** [review]
good idea and thanks for taking the initiative! :shipit:

**Pull request merged**